### ISSUE TITLE
replace_nans_at_start overwrite existing values

### ISSUE THREAD

**Issue by JadM133** [patrick-kidger/diffrax#324]
Hello, I am not sure if the behavior of **replace_nans_at_start** in **backward_hermite_coefficients** is expected when we have initial values that aren't all NaNs.

The documentation proposes that "replace_nans_at_start" will change the NaN values at the start by the ones given.
 
I believe that this parameter is only useful if **ALL** the values at the start are NaNs. I give an example below (based again on get_data from the NCDE example) where the dataset has **no** NaN values. It actually raises an assertion error because values are being switched to "means" regardless wether they're NaNs or not.

![snap](https://github.com/patrick-kidger/diffrax/assets/123482935/d0a69ecc-593e-4538-a65a-d7f685e4441d)

I think the problem is that in **_backward_hermite_coefficients** the following line:

**y0 = jnp.broadcast_to(replace_nans_at_start, ys[0].shape)**

doesn't consider if the values are NaNs and switches them automatically. In my opinion, the logic behind the parameter would be to only switch the NaN values and not overwrite the already existing ones. I would propose either mentioning in the docstring if this is the expected behvaior or adjusting the proposed line to consider if values are actually NaNs or not.

Thank you in advance! The library is great!

**Comment by patrick-kidger** [comment]
Yup, agreed. This at least should be an easy fix -- if you get the chance I'd be happy to take a pull request on this!
